### PR TITLE
Root based filesystem support for FTP server

### LINKED ISSUE #1409: Cannot browse /data/data via FTP even though Settings > Advanced > Root Explorer is active
I activated Settings > Advanced > Root Explorer and am able to browse /data/data just fine in the regular File Manager. However, the FTP Server doesn't work with this path. I am using FileZilla to connect to the server and no files or folders are listed.

Thank you very much in advance. :+1:

### PR DESCRIPTION
**Pull request by TranceLove**
## Description
- Root-based file system backend for Apache ftpserver backed by libsu's SuFile
- Logic to switch to root based filesystem if user selects the sharing folder is outside storage dirs which doesn't require root
- Warns user of using this root based filesystem implementation

Test case for the file system implementation cannot be done due to shadowing SuFile was not possible with unknown reasons.
- `ls -a` on normal UNIXes behaves differently than busybox's `ls -a`, causing the need of shadowing SuFile
- however test cases using SuFile shadow would fail with weird exceptions related to Robolectric's bytecode manipulation

There is one known issue though: when choosing the sharing folder, browsing above storage level directories may not work since they require root access. When using on Oneplus 2 with LineageOS 18, no directories can be seen above storage level.
(May be solved along with #3061?)

#### Issue tracker   
Fixes #1409

#### Automatic tests
- [ ] Added test cases
  
#### Manual tests
- [x] Done  
  
Devices:
- GPD XD running LegacyROM (4.4.4)
- Oneplus 2 running LineageOS 18 (11)

Browsing `/data/data` or protected directories like `Android/data` shall function normally.

#### Build tasks success  
Successfully running following tasks on local:
- [x] `./gradlew assembledebug`
- [x] `./gradlew spotlessCheck`

### CODE REVIEW

File: app/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -717,7 +717,9 @@ You only need to do this once, until the next time you select a new location for
     <string name="android_data_prompt_saf_access_title">Grant SAF access</string>
     <string name="ftp_preference_saf_filesystem_title">Use legacy filesystem</string>
     <string name="ftp_preference_saf_filesystem">Enable this for support for using device external storage on newer Android versions</string>
-    <string name="ftp_server_fallback_path_reset_prompt">FTP server shared path had been resetted to internal storage as you are switching back to legacy filesystem implementation. Please select a new path using the menu on the top right as necessary.</string>
+    <string name="ftp_server_fallback_path_reset_prompt">FTP server shared path had been reset to internal storage as you are switching back to legacy filesystem implementation. Please select a new path using the menu on the top right as necessary.</string>\
```
Comment (EmmanuelMess) [inline app/src/main/res/values/strings.xml:720]: What's up with the slash at the end?

File: app/src/main/java/com/amaze/filemanager/crashreport/ErrorActivity.java [hunk 2]
```diff
@@ -147,7 +147,13 @@ public static void reportError(
       final Context context, final CrashReportData report, final ErrorInfo errorInfo) {
     System.out.println("ErrorActivity reportError");
     final String[] el = new String[] {report.getString(ReportField.STACK_TRACE)};
-
+    // Add this to try figure out what happened when stacktrace is sent to acra.
+    // Hope this will be useful for build failures...
+    if (BuildConfig.DEBUG) {
+      for(String line: el) {
+        System.out.println(line);
```
Comment (EmmanuelMess) [inline app/src/main/java/com/amaze/filemanager/crashreport/ErrorActivity.java:154]: `System.out.print` should not be used in Android. Please use standard log, or add the info directly into the acra report.

File: app/src/main/java/com/amaze/filemanager/filesystem/ftpserver/RootFileSystemView.kt [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+import java.util.*
+
+class RootFileSystemView(private val user: User) : FileSystemView {
+
+    private var currDir: String
+    private var rootDir: String
+
+    companion object {
+        @JvmStatic
+        private val logger: Logger = LoggerFactory.getLogger(RootFileSystemView::class.java)
```
Comment (EmmanuelMess) [inline app/src/main/java/com/amaze/filemanager/filesystem/ftpserver/RootFileSystemView.kt:19]: Why are we using these loggers? Aren't android's good?

File: app/src/main/java/com/amaze/filemanager/ui/dialogs/GeneralDialogCreation.java [hunk 4]
```diff
@@ -149,6 +149,25 @@ public static MaterialDialog showBasicDialog(
     return a.build();
   }
 
+  public static MaterialDialog createAlertDialog(
```
Comment (EmmanuelMess) [inline app/src/main/java/com/amaze/filemanager/ui/dialogs/GeneralDialogCreation.java:152]: Please move this to a kt object in dialogs package, annotated with `@JvmStatic`. Like dialogs in  #3205.

**Review by EmmanuelMess — CHANGES_REQUESTED** [review]
How do I test this? Is it possible to create tests for this?

**Comment by EmmanuelMess** [comment]
How do I test if the PR works?

**Comment by EmmanuelMess** [comment]
I am having this error with any folder I set, so I cant test :sweat_smile: 

```
2022-03-08 05:59:24.039 2752-2885/? W/System.err: [NioProcessor-1] WARN org.apache.mina.util.DefaultExceptionMonitor - Unexpected exception.
2022-03-08 05:59:24.039 2752-2885/? W/System.err: org.apache.mina.core.filterchain.IoFilterLifeCycleException: onPreAdd(): sslFilter:SslFilter in (0x00000004: nio socket, server, /192.168.0.85:45176 => /192.168.0.169:2211)
2022-03-08 05:59:24.040 2752-2885/? W/System.err:     at org.apache.mina.core.filterchain.DefaultIoFilterChain.register(DefaultIoFilterChain.java:383)
2022-03-08 05:59:24.041 2752-2885/? W/System.err:     at org.apache.mina.core.filterchain.DefaultIoFilterChain.addLast(DefaultIoFilterChain.java:189)
2022-03-08 05:59:24.043 2752-2885/? W/System.err:     at org.apache.mina.core.filterchain.DefaultIoFilterChainBuilder.buildFilterChain(DefaultIoFilterChainBuilder.java:496)
2022-03-08 05:59:24.043 2752-2885/? W/System.err:     at org.apache.mina.core.polling.AbstractPollingIoProcessor.addNow(AbstractPollingIoProcessor.java:512)
2022-03-08 05:59:24.044 2752-2885/? W/System.err:     at org.apache.mina.core.polling.AbstractPollingIoProcessor.handleNewSessions(AbstractPollingIoProcessor.java:487)
2022-03-08 05:59:24.045 2752-2885/? W/System.err:     at org.apache.mina.core.polling.AbstractPollingIoProcessor.access$400(AbstractPollingIoProcessor.java:68)
2022-03-08 05:59:24.045 2752-2885/? W/System.err:     at org.apache.mina.core.polling.AbstractPollingIoProcessor$Processor.run(AbstractPollingIoProcessor.java:1111)
2022-03-08 05:59:24.046 2752-2885/? W/System.err:     at org.apache.mina.util.NamePreservingRunnable.run(NamePreservingRunnable.java:64)
2022-03-08 05:59:24.046 2752-2885/? W/System.err:     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1076)
2022-03-08 05:59:24.046 2752-2885/? W/System.err:     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:569)
2022-03-08 05:59:24.047 2752-2885/? W/System.err:     at java.lang.Thread.run(Thread.java:856)
2022-03-08 05:59:24.048 2752-2885/? W/System.err: Caused by: java.lang.IllegalArgumentException: TLS_ECDH_ECDSA_WITH_RC4_128_SHA is not supported.
2022-03-08 05:59:24.049 2752-2885/? W/System.err:     at org.apache.harmony.xnet.provider.jsse.SSLParametersImpl.setEnabledCipherSuites(SSLParametersImpl.java:238)
2022-03-08 05:59:24.049 2752-2885/? W/System.err:     at org.apache.harmony.xnet.provider.jsse.SSLEngineImpl.setEnabledCipherSuites(SSLEngineImpl.java:221)
2022-03-08 05:59:24.050 2752-2885/? W/System.err:     at org.apache.mina.filter.ssl.SslHandler.init(SslHandler.java:175)
2022-03-08 05:59:24.050 2752-2885/? W/System.err:     at org.apache.mina.filter.ssl.SslFilter.onPreAdd(SslFilter.java:445)
2022-03-08 05:59:24.052 2752-2885/? W/System.err:     at org.apache.mina.core.filterchain.DefaultIoFilterChain.register(DefaultIoFilterChain.java:381)
2022-03-08 05:59:24.053 2752-2885/? W/System.err: 	... 10 more
2022-03-08 05:59:29.245 2752-2885/? W/System.err: [NioProcessor-1] WARN org.apache.mina.util.DefaultExceptionMonitor - Unexpected exception.
2022-03-08 05:59:29.245 2752-2885/? W/System.err: org.apache.mina.core.filterchain.IoFilterLifeCycleException: onPreAdd(): sslFilter:SslFilter in (0x00000005: nio socket, server, /192.168.0.85:45180 => /192.168.0.169:2211)
2022-03-08 05:59:29.247 2752-2885/? W/System.err:     at org.apache.mina.core.filterchain.DefaultIoFilterChain.register(DefaultIoFilterChain.java:383)
2022-03-08 05:59:29.248 2752-2885/? W/System.err:     at org.apache.mina.core.filterchain.DefaultIoFilterChain.addLast(DefaultIoFilterChain.java:189)
2022-03-08 05:59:29.248 2752-2885/? W/System.err:     at org.apache.mina.core.filterchain.DefaultIoFilterChainBuilder.buildFilterChain(DefaultIoFilterChainBuilder.java:496)
2022-03-08 05:59:29.248 2752-2885/? W/System.err:     at org.apache.mina.core.polling.AbstractPollingIoProcessor.addNow(AbstractPollingIoProcessor.java:512)
2022-03-08 05:59:29.248 2752-2885/? W/System.err:     at org.apache.mina.core.polling.AbstractPollingIoProcessor.handleNewSessions(AbstractPollingIoProcessor.java:487)
2022-03-08 05:59:29.248 2752-2885/? W/System.err:     at org.apache.mina.core.polling.AbstractPollingIoProcessor.access$400(AbstractPollingIoProcessor.java:68)
2022-03-08 05:59:29.249 2752-2885/? W/System.err:     at org.apache.mina.core.polling.AbstractPollingIoProcessor$Processor.run(AbstractPollingIoProcessor.java:1111)
2022-03-08 05:59:29.249 2752-2885/? W/System.err:     at org.apache.mina.util.NamePreservingRunnable.run(NamePreservingRunnable.java:64)
2022-03-08 05:59:29.250 2752-2885/? W/System.err:     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1076)
2022-03-08 05:59:29.250 2752-2885/? W/System.err:     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:569)
2022-03-08 05:59:29.250 2752-2885/? W/System.err:     at java.lang.Thread.run(Thread.java:856)
2022-03-08 05:59:29.250 2752-2885/? W/System.err: Caused by: java.lang.IllegalArgumentException: TLS_ECDH_ECDSA_WITH_RC4_128_SHA is not supported.
2022-03-08 05:59:29.252 2752-2885/? W/System.err:     at org.apache.harmony.xnet.provider.jsse.SSLParametersImpl.setEnabledCipherSuites(SSLParametersImpl.java:238)
2022-03-08 05:59:29.252 2752-2885/? W/System.err:     at org.apache.harmony.xnet.provider.jsse.SSLEngineImpl.setEnabledCipherSuites(SSLEngineImpl.java:221)
2022-03-08 05:59:29.253 2752-2885/? W/System.err:     at org.apache.mina.filter.ssl.SslHandler.init(SslHandler.java:175)
2022-03-08 05:59:29.254 2752-2885/? W/System.err:     at org.apache.mina.filter.ssl.SslFilter.onPreAdd(SslFilter.java:445)
2022-03-08 05:59:29.254 2752-2885/? W/System.err:     at org.apache.mina.core.filterchain.DefaultIoFilterChain.register(DefaultIoFilterChain.java:381)
2022-03-08 05:59:29.255 2752-2885/? W/System.err: 	... 10 more
```

**Comment by TranceLove** [comment]
Seems to me you are using [a lower API device](https://developer.android.com/reference/javax/net/ssl/SSLEngine#default-configuration-for-different-android-versions). Never mind, will prepare tests for that.

**Comment by EmmanuelMess** [comment]
Yes

**Review by VishalNehra — APPROVED** [review]

**Pull request merged**